web/api/badges/web_buffer_svg.c

```diff
@@ -139,153 +139,35 @@ static double verdana11_widths[256] = {
     [124] = 4.9951171875, // |
     [125] = 6.982421875, // }
     [126] = 9.001953125, // ~
-    [127] = 0.0,
-    [128] = 0.0,
-    [129] = 0.0,
-    [130] = 0.0,
-    [131] = 0.0,
-    [132] = 0.0,
-    [133] = 0.0,
-    [134] = 0.0,
-    [135] = 0.0,
-    [136] = 0.0,
-    [137] = 0.0,
-    [138] = 0.0,
-    [139] = 0.0,
-    [140] = 0.0,
-    [141] = 0.0,
-    [142] = 0.0,
-    [143] = 0.0,
-    [144] = 0.0,
-    [145] = 0.0,
-    [146] = 0.0,
-    [147] = 0.0,
-    [148] = 0.0,
-    [149] = 0.0,
-    [150] = 0.0,
-    [151] = 0.0,
-    [152] = 0.0,
-    [153] = 0.0,
-    [154] = 0.0,
-    [155] = 0.0,
-    [156] = 0.0,
-    [157] = 0.0,
-    [158] = 0.0,
-    [159] = 0.0,
-    [160] = 0.0,
-    [161] = 0.0,
-    [162] = 0.0,
-    [163] = 0.0,
-    [164] = 0.0,
-    [165] = 0.0,
-    [166] = 0.0,
-    [167] = 0.0,
-    [168] = 0.0,
-    [169] = 0.0,
-    [170] = 0.0,
-    [171] = 0.0,
-    [172] = 0.0,
-    [173] = 0.0,
-    [174] = 0.0,
-    [175] = 0.0,
-    [176] = 0.0,
-    [177] = 0.0,
-    [178] = 0.0,
-    [179] = 0.0,
-    [180] = 0.0,
-    [181] = 0.0,
-    [182] = 0.0,
-    [183] = 0.0,
-    [184] = 0.0,
-    [185] = 0.0,
-    [186] = 0.0,
-    [187] = 0.0,
-    [188] = 0.0,
-    [189] = 0.0,
-    [190] = 0.0,
-    [191] = 0.0,
-    [192] = 0.0,
-    [193] = 0.0,
-    [194] = 0.0,
-    [195] = 0.0,
-    [196] = 0.0,
-    [197] = 0.0,
-    [198] = 0.0,
-    [199] = 0.0,
-    [200] = 0.0,
-    [201] = 0.0,
-    [202] = 0.0,
-    [203] = 0.0,
-    [204] = 0.0,
-    [205] = 0.0,
-    [206] = 0.0,
-    [207] = 0.0,
-    [208] = 0.0,
-    [209] = 0.0,
-    [210] = 0.0,
-    [211] = 0.0,
-    [212] = 0.0,
-    [213] = 0.0,
-    [214] = 0.0,
-    [215] = 0.0,
-    [216] = 0.0,
-    [217] = 0.0,
-    [218] = 0.0,
-    [219] = 0.0,
-    [220] = 0.0,
-    [221] = 0.0,
-    [222] = 0.0,
-    [223] = 0.0,
-    [224] = 0.0,
-    [225] = 0.0,
-    [226] = 0.0,
-    [227] = 0.0,
-    [228] = 0.0,
-    [229] = 0.0,
-    [230] = 0.0,
-    [231] = 0.0,
-    [232] = 0.0,
-    [233] = 0.0,
-    [234] = 0.0,
-    [235] = 0.0,
-    [236] = 0.0,
-    [237] = 0.0,
-    [238] = 0.0,
-    [239] = 0.0,
-    [240] = 0.0,
-    [241] = 0.0,
-    [242] = 0.0,
-    [243] = 0.0,
-    [244] = 0.0,
-    [245] = 0.0,
-    [246] = 0.0,
-    [247] = 0.0,
-    [248] = 0.0,
-    [249] = 0.0,
-    [250] = 0.0,
-    [251] = 0.0,
-    [252] = 0.0,
-    [253] = 0.0,
-    [254] = 0.0,
-    [255] = 0.0
```

Comment:
Hi there are 2 parts to this at least so it seems to me at first glance:

1. URL parser %c2 issues -> will check that and come back to you. Today very busy with other things.
I know some parts of URL parser clearly needs improving. I do not like the old code which kind of tries to replace unknowns with spaces... IMHO this is dangerous and we need to have a way to "fail parsing" and return HTTP code 400 to client as soon as we see something suspicious. But as i mentioned I do not want to put everything into one huge pull request.
2. verdana11_width -> here I am confused maybe what you mean
With newest changes verdana11_width it doesn't modify the string and just returns width of it (best guess basis, as later it is updated by javascript)
```` C
if(likely(!(*s & 0x80))){ // Byte 1XXX XXXX is not valid in UTF8
     double t = verdana11_widths[(unsigned char)*s];
````
Here we check if the value is < 128 and only then we access this array (bound check). Your complaint is that we access array of size 128 with char which is -128 to 127 (0-255)? But wouldn't following be the same case?:
```` C
#define SAMPLE_ARRY_SIZE 10
int sample_arry[SAMPLE_ARRY_SIZE];
for (int i = 0; i < SAMPLE_ARRY_SIZE; i++){
   sample_arry[i] = whatever;
} 
````
You are accessing array of size 10 with integer which can be huge. That is however normal thing as long as you bound check before (which in C is responsibility of developer). Same you do with new verdana11_width code (there is bound check `*s & 0x80`).
Or is your point that you want to make that bound check more obvious with `#define`?

Reply:
About 1, it is very good idea to concentrate in the problem instead the whole url parser, it was already rewritten and it will enter in test soon. We also know about answer problems that we have with it, and in the correct time it will be fixed. Thanks to remember us this. When you finish to solve the problem with %c2 let us know!
About 2 and the (unsigned char ) cast that follows,  I did some math converting the hex 0x80(128) to the binary and I agree with you, it will never go nothing bigger and equal than 128 due this conditional.
I won't mark this talk as resolved, because you cited the bug '%c2' here, but the rest is ok.

Reply:
Hi, so I am able to get to it now. Credits to my wife for caring for the small one.

### %82
So when I sent http://127.0.0.1:19999/api/v1/badge.svg?chart=system.cpu&label=%82
I get:
![82](https://user-images.githubusercontent.com/6674623/61168746-ffdbb180-a552-11e9-8c68-bf6686a11586.png)

Reason: %82 is invalid character in UTF-8 therefore we should reject it.
Good way to handle it: return 400 to client and do not attempt to parse further for security reasons. Maybe create issue for that and fix it as next step
**How is it handled now:**
%82 is recognized as invalid and % is skipped and "82" is then taken as normal string (not prepended by 82)

Original code from TAG v1.16.0:
````C
                char t = from_hex(s[1]) << 4 | from_hex(s[2]);
                // avoid HTTP header injection
                *d++ = (char)((isprint(t))? t : ' ');
                s += 2;
````
will skip also those 2 chars which I also can do. However as I mentioned in new code:
````C
                else {
                    //TODO!: Should stop parsing and send HTTP error code 400 back to client instead
                    //      same goes for other errors - if we see something suspicious abort and send 400
                    *d++ = ' ';
````

### %c2
I cannot reproduce the error you mention. With latest version I committed it behaves completely same as %82:
![c2](https://user-images.githubusercontent.com/6674623/61168888-feab8400-a554-11e9-9e8b-298ff43b31c4.png)
Here the slight difference being that %c2 (1100 0010b) is first byte of multibyte character but there is no second byte... So at least at my PC with latest commit applied it correctly recognizes that this is not valid UTF8 character and rejects it by ignoring % and handling chars behind it as ASCII same as in case of %82.

Therefore IMHO same applies as before: best would be to return HTTP code 400 and immediately stop parsing (as URL is clearly malformed - maybe somebody is trying funky stuff etc.). Current solution is just ignore % char if there is something wrong.

>Unfortunately i can't reproduce the:
>XML Parsing Error: not well-formed
>Location: http://192.168.0.17:19999/api/v1/badge.svg?chart=system.cpu&label=%c2
>Line Number 1, Column 819:

Reply:
Hi, so i took a look at the code and it's not that hard to do 400 error return instead. So in latest commit I implemented it. Now if not valid utf8 found we return 400 BAD REQUEST